consensus/src/ledger/mod.rs

```diff
@@ -0,0 +1,238 @@
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+    path::Path,
+    sync::Arc,
+};
+
+use anyhow::*;
+use dyn_clone::DynClone;
+use smallvec::SmallVec;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::{
+    merkle_tree::{MerklePath, MerkleTreeDigest},
+    CommitmentScheme,
+    MerkleParameters,
+    SignatureScheme,
+};
+use snarkvm_dpc::{
+    testnet1::{BaseDPCComponents, Transaction},
+    Block,
+    LedgerScheme,
+    TransactionScheme,
+};
+
+mod merkle;
+pub use merkle::MerkleLedger;
+mod indexed_merkle_tree;
+pub use indexed_merkle_tree::IndexedMerkleTree;
+use snarkvm_parameters::{LedgerMerkleTreeParameters, Parameter};
+use snarkvm_utilities::{FromBytes, ToBytes};
+
+pub trait Ledger: Send + Sync + DynClone {
+    fn extend(&mut self, new_commitments: &[Digest], new_serials: &[Digest], new_memos: &[Digest]) -> Result<Digest>;
+
+    fn rollback(&mut self, commitments: &[Digest], serials: &[Digest], memos: &[Digest]) -> Result<()>;
+
+    fn commitment_len(&self) -> usize;
+
+    fn contains_commitment(&self, commitment: &Digest) -> bool;
+
+    fn commitment_index(&self, commitment: &Digest) -> Option<usize>;
+
+    fn contains_serial(&self, serial: &Digest) -> bool;
+
+    fn contains_memo(&self, memo: &Digest) -> bool;
+
+    fn validate_digest(&self, digest: &Digest) -> bool;
+
+    fn digest(&self) -> Digest;
+
+    fn generate_proof(&self, commitment: &Digest, index: usize) -> Result<Vec<(Digest, Digest)>>;
+
+    /// checks if a ledgers state is consistent
+    fn validate_ledger(&self) -> bool;
+}
+
+pub struct DynLedger(pub Box<dyn Ledger>);
+
+impl Clone for DynLedger {
+    fn clone(&self) -> Self {
+        DynLedger(dyn_clone::clone_box(&*self.0))
+    }
+}
+
+impl Deref for DynLedger {
+    type Target = dyn Ledger;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.0
+    }
+}
+
+impl DerefMut for DynLedger {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.0
+    }
+}
+
+pub struct DeserializedLedger<'a, C: BaseDPCComponents> {
+    parameters: Arc<C::MerkleParameters>,
+    inner: &'a DynLedger,
+    _p: PhantomData<C>,
+}
+
+impl DynLedger {
+    pub fn deserialize<'a, C: BaseDPCComponents>(&'a self) -> DeserializedLedger<'a, C> {
+        //todo: cache this
+        let crh = <C::MerkleParameters as MerkleParameters>::H::from(
+            FromBytes::read(&LedgerMerkleTreeParameters::load_bytes().expect("failed to load merkle parameters")[..])
+                .expect("failed to read merkle parameters"),
+        );
+        let parameters = Arc::new(C::MerkleParameters::from(crh));
+
+        DeserializedLedger {
+            parameters,
+            inner: self,
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<'a, C: BaseDPCComponents> DeserializedLedger<'a, C> {
+    pub fn serialize(self) -> &'a DynLedger {
```

Comment:
This name is misleading - this just returns the trait object; also, is the `DynLedger` sent through some channel anywhere? I didn't see it happen, so why use a trait object here?

Reply:
to shield type parameters and allow different implementations. it's not misleading? it is returning the interior 'serialized' context ledger reference.

Reply:
Perhaps I've been going over the snarkVM codebase for too long recently, but (de)serialization heavily implies going from/to raw bytes there, and that might be confusing.